### PR TITLE
PVP-168

### PR DESCRIPTION
**Pull request by vytjar**
Added processing of ExerciseSessionRecord (HealthConnect data type)

- Implemented distance reading from Healthconnect
- Created a new model - ExerciseSessionInfo that stored a record and its duration, distance (information will probably be expanded in the future)
- Added new service that creates ExerciseSessionInfo from ExerciseSessionRecord
- Also added a method for converting list of ExerciseSessionRecord to list of SportActivity at Utils.kt
- Added method to calculate occurences of each sport activity in a list

### CODE REVIEW

File: app/src/main/java/com/pvp/app/api/ExerciseService.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.pvp.app.api
+
+import androidx.health.connect.client.records.ExerciseSessionRecord
+import com.pvp.app.model.ExerciseSessionInfo
+
+interface ExerciseService {
+    /**
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/api/ExerciseService.kt:7]: Separation.

File: app/src/main/java/com/pvp/app/api/HealthConnectService.kt [hunk 2]
```diff
@@ -29,4 +29,16 @@ interface HealthConnectService {
...
+
+    /**
+     * Aggregates the distance user has traversed (by sport activities)
+     * between the specified time range
+     * @param start Specifies the start of time range
+     * @param end Specifies the end of the time range
+     * @return Returns the distance in meters
+     */
+    suspend fun aggregateDistance(
+        start: java.time.Instant,
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/api/HealthConnectService.kt:41]: Could we import `java.time.Instant` into the file to evade such long type declarations?

File: app/src/main/java/com/pvp/app/common/Utils.kt [hunk 3]
```diff
@@ -44,4 +46,17 @@ fun Picture.toImageBitmap(): ImageBitmap {
...
+}
+
+fun List<ExerciseSessionRecord>.toSportActivities(): List<SportActivity> {
+    return this.mapNotNull { exercise ->
+        SportActivity.fromId(exercise.exerciseType)
+    }
+}
+
+fun List<SportActivity>.getOccurences(): List<Pair<SportActivity, Int>> {
+    return this.groupingBy { it }
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/common/Utils.kt:58]: Split.

File: app/src/main/java/com/pvp/app/di/ServiceModule.kt [hunk 4]
```diff
@@ -51,4 +53,8 @@ interface ServiceModule {
     @Binds
     @Singleton
     fun bindUserService(service: UserServiceImpl): UserService
+
+    @Binds
+    @Singleton
+    fun bindExerciseService(service: ExerciseServiceImpl): ExerciseService
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/di/ServiceModule.kt:59]: Could this be put after `bindConfiguration` to keep the alphabet order?

File: app/src/main/java/com/pvp/app/service/ExerciseServiceImpl.kt [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+    override suspend fun getExerciseInfo(record: ExerciseSessionRecord): ExerciseSessionInfo {
+        return ExerciseSessionInfo(
+            record = record,
+            distance = when (SportActivity.fromId(record.exerciseType)?.supportsDistanceMetrics) {
+                true -> service.aggregateDistance(record.startTime, record.endTime)
+                else -> null
+            },
+            duration = Duration.between(record.startTime, record.endTime)
+        )
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/service/ExerciseServiceImpl.kt:23]:
```kotlin
        return ExerciseSessionInfo(
            record = record,
            distance = when (SportActivity.fromId(record.exerciseType)?.supportsDistanceMetrics) {
                true -> service.aggregateDistance(
                    record.startTime,
                    record.endTime
                )
                else -> null
            },
            duration = Duration.between(
                record.startTime,
                record.endTime
            )
        )
```

File: app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt [hunk 6]
```diff
@@ -44,4 +47,22 @@ class HealthConnectServiceImpl @Inject constructor(
...
+
+    override suspend fun aggregateDistance(
+        start: Instant,
+        end: Instant
+    ): Double {
+        return try {
+            val response = client.aggregate(
+                AggregateRequest(
+                    metrics = setOf(DistanceRecord.DISTANCE_TOTAL),
+                    timeRangeFilter = TimeRangeFilter.between(start, end)
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt:59]: Split params.

File: app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt [hunk 7]
```diff
@@ -44,4 +47,22 @@ class HealthConnectServiceImpl @Inject constructor(
...
+                )
+            )
+            response[DistanceRecord.DISTANCE_TOTAL]?.inMeters ?: 0.0
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt:62]: Separate return from variable block.

File: app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt [hunk 8]
```diff
@@ -44,4 +47,22 @@ class HealthConnectServiceImpl @Inject constructor(
...
+        } catch (e: Exception) {
+            Log.e("DISTANCE", "EXCEPTION ${e.toString()}")
+            0.0
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt:65]: Same as above.

File: app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt [hunk 9]
```diff
@@ -44,4 +47,22 @@ class HealthConnectServiceImpl @Inject constructor(
...
+        return try {
+            val response = client.aggregate(
+                AggregateRequest(
+                    metrics = setOf(DistanceRecord.DISTANCE_TOTAL),
+                    timeRangeFilter = TimeRangeFilter.between(start, end)
+                )
+            )
+            response[DistanceRecord.DISTANCE_TOTAL]?.inMeters ?: 0.0
+        } catch (e: Exception) {
+            Log.e("DISTANCE", "EXCEPTION ${e.toString()}")
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/service/HealthConnectServiceImpl.kt:64]: `{e}` would automatically be converted to string when used in a string interpolation/concatation, if I recall correctly. Or if a message is all you need, use `${e.message}` then.

File: app/src/main/java/com/pvp/app/ui/screen/steps/StepScreen.kt [hunk 10]
```diff
@@ -45,13 +46,8 @@ fun ActivityList(
...
 
-    val activities = model.activites
-        .collectAsStateWithLifecycle().value
-        .groupingBy { it.title }
-        .eachCount()
-        .toList()
-        .sortedByDescending { it.second }
-
+    val activities = model.activities
+        .collectAsStateWithLifecycle().value.getOccurences()
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/steps/StepScreen.kt:50]: Split after value.

File: app/src/main/java/com/pvp/app/ui/screen/steps/StepViewModel.kt [hunk 11]
```diff
@@ -74,14 +75,12 @@ class StepViewModel @Inject constructor(
...
-            }
+            _activities.value = records.toSportActivities()
         }
     }
 }
 
 val PERMISSIONS = setOf(
     HealthPermission.getReadPermission(StepsRecord::class),
-    HealthPermission.getReadPermission(ExerciseSessionRecord::class)
+    HealthPermission.getReadPermission(ExerciseSessionRecord::class),
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/ui/screen/steps/StepViewModel.kt:85]: Seems redundant change.

**Review by klaidoshka — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/pvp/app/di/ServiceModule.kt [hunk 12]
```diff
@@ -32,6 +34,10 @@ interface ServiceModule {
     @Singleton
     fun bindAuthenticationService(service: AuthenticationServiceImpl): AuthenticationService
 
+    @Binds
+    @Singleton
+    fun bindExerciseService(service: ExerciseServiceImpl): ExerciseService
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/di/ServiceModule.kt:39]: ABCDE, it should be after `bindConfiguration`.

File: app/src/main/java/com/pvp/app/common/Utils.kt [hunk 13]
```diff
@@ -44,4 +46,18 @@ fun Picture.toImageBitmap(): ImageBitmap {
...
+}
+
+fun List<ExerciseSessionRecord>.toSportActivities(): List<SportActivity> {
+    return this.mapNotNull { exercise ->
+        SportActivity.fromId(exercise.exerciseType)
+    }
+}
+
+fun List<SportActivity>.getOccurences(): List<Pair<SportActivity, Int>> {
+    return this
```
Comment (klaidoshka) [inline app/src/main/java/com/pvp/app/common/Utils.kt:58]: Also note that you are in `List<SportActivity>` instance context, so you can call all methods of the list without specifying `this`. If you want to use it, let it be. If you don't want to, then it is possible, just call it as so:
```kotlin
    return groupingBy { it }
        .eachCount()
        .toList()
        .sortedByDescending { it.second }
```

**Review by klaidoshka — APPROVED** [review]

File: app/src/main/java/com/pvp/app/di/ServiceModule.kt (hunk 12 above)
Comment (vytjar) [reply app/src/main/java/com/pvp/app/di/ServiceModule.kt:39]: Yup gonna relearn the alphabet, mb

**Review by KhajiitLikeToSneak — APPROVED** [review]

**Pull request merged**